x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/DataStreamDeprecationChecks.java

```diff
@@ -14,61 +14,37 @@
 import org.elasticsearch.xpack.core.deprecation.DeprecationIssue;
 
 import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
 
 import static java.util.Map.entry;
 import static java.util.Map.ofEntries;
 
 public class DataStreamDeprecationChecks {
     static DeprecationIssue oldIndicesCheck(DataStream dataStream, ClusterState clusterState) {
         List<Index> backingIndices = dataStream.getIndices();
-        boolean hasOldIndices = backingIndices.stream()
-            .anyMatch(index -> clusterState.metadata().index(index).getCompatibilityVersion().before(IndexVersions.V_8_0_0));
-        if (hasOldIndices) {
-            long totalIndices = backingIndices.size();
-            List<Index> oldIndices = backingIndices.stream()
-                .filter(index -> clusterState.metadata().index(index).getCompatibilityVersion().before(IndexVersions.V_8_0_0))
-                .toList();
-            long totalOldIndices = oldIndices.size();
-            long totalOldSearchableSnapshots = oldIndices.stream()
-                .filter(index -> clusterState.metadata().index(index).isSearchableSnapshot())
-                .count();
-            long totalOldPartiallyMountedSearchableSnapshots = oldIndices.stream()
-                .filter(index -> clusterState.metadata().index(index).isPartialSearchableSnapshot())
-                .count();
-            long totalOldFullyMountedSearchableSnapshots = totalOldSearchableSnapshots - totalOldPartiallyMountedSearchableSnapshots;
+
+        Set<String> indiciesNeedingUpgrade = backingIndices.stream()
+            .filter(index -> clusterState.metadata().index(index).getCompatibilityVersion().before(IndexVersions.V_8_0_0))
```

Comment:
Perhaps this index version filtering logic should be the same as: https://github.com/elastic/elasticsearch/blob/0a6ce27825d15a7e8294a99ee1aca52eb05f8ae0/x-pack/plugin/migrate/src/main/java/org/elasticsearch/xpack/migrate/action/ReindexDataStreamAction.java#L61

I'm not sure what the difference is between compatibility version and created version, but we should make sure they are consistent.

Reply:
Yeah I think it probably ought to use the predicate that's in that action, because I suspect we'll be adding things to it (like skipping indices that are marked as being ok to be read only). We can move that predicate if we need to.

Reply:
I'm not so sure this is right now... The text of the deprecation message says `his data stream has backing indices that were created before Elasticsearch 8.0.0` but this predicate copied from the migrate API actually would flag any index created in 9.0 or below...

Reply:
Yeah in main i will be indices created before 9.0.0, and when we backport to 8.x it will be before 8.0.0. I think that was a bug in wording that I introduced when I initially did it.

Reply:
> I could move the predicate into a common module (server) but wasn't sure on where exactly and if it's a good idea...

Yeah I don't know that I'd want to move it into server. Maybe x-pack/core? We can do it in a subsequent PR. But you'll want to make sure to also update the version when you backport the same way the other predicate does.

Reply:
OK thanks for the clarification. I'll get that wording fixed and see about moving the predicate :-)